changefeedccl: support for constant headers for webhook and kafka sinks

asg0451 opened the pull request:
This adds support for constant headers for webhook
and kafka sinks, to be specified in the CREATE
CHANGEFEED statement as a JSON object.

Part of: #142273
Epic: CRDB-48880

Release note (general change): The CREATE
CHANGEFEED statement now supports the
`extra_headers` option, which can be used to
specify extra headers for webhook and kafka
sinks. This can be used to add headers to all
messages sent to the sink.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146813)
<!-- Reviewable:end -->

asg0451 committed:
changefeedccl: support for constant headers for webhook and kafka sinks

This adds support for constant headers for webhook
and kafka sinks, to be specified in the CREATE
CHANGEFEED statement as a JSON object.

Part of: #142273
Epic: CRDB-48880

Release note (general change): The CREATE
CHANGEFEED statement now supports the
`extra_headers` option, which can be used to
specify extra headers for webhook and kafka
sinks. This can be used to add headers to all
messages sent to the sink.

KeithCh commented on pkg/ccl/changefeedccl/cdctest/testfeed.go:
```diff
@@ -39,7 +41,10 @@ func (h Headers) String() string {
+	sorted := slices.SortedFunc(slices.Values(h), func(a, b Header) int {
+		return strings.Compare(a.K, b.K)
+	})
+	for _, v := range sorted {
```
What's the sorting for?

andyyang890 commented on pkg/ccl/changefeedccl/changefeedbase/options.go:
```diff
@@ -1049,13 +1054,66 @@ func (s StatementOptions) GetWebhookSinkOptions() (WebhookSinkOptions, error) {
+	Headers map[string][]byte
+}
+
+func (s StatementOptions) GetKafkaSinkOptions() (KafkaSinkOptions, error) {
```
nit: seems like a bit of a misnomer given it's also used by pulsar, maybe `SinkSpecificOptions` instead?

asg0451 replied:
i dont think changing that is in scope for this PR. i've never even read that code and intend to keep it that way ;)

asg0451 commented on pkg/ccl/changefeedccl/changefeedbase/options.go:
```diff
@@ -1049,13 +1054,66 @@ func (s StatementOptions) GetWebhookSinkOptions() (WebhookSinkOptions, error) {
-// by the kafka sink.
-func (s StatementOptions) GetKafkaConfigJSON() SinkSpecificJSONConfig {
-	return s.getJSONValue(OptKafkaSinkConfig)
+func parseHeaders[RT interface{ string | []byte }](headers string) (map[string]RT, error) {
```
is this a standard we've established or just your preference?

asg0451 commented on pkg/ccl/changefeedccl/changefeedbase/options.go:
```diff
@@ -117,6 +118,7 @@ const (
 	// TODO(#142273): look into whether we want to add headers to pub/sub, and other
 	// sinks as well (eg cloudstorage, webhook, ..). Currently it's kafka-only.
 	OptHeadersJSONColumnName = `headers_json_column_name`
+	OptHeaders               = `headers`
```
going with extra_headers

asg0451 commented:
bors r=andyyang890,KeithCh
